plugin/pkg/shoot/validator/admission.go

```diff
@@ -366,32 +366,49 @@ func (c *validationContext) validateSeedSelectionForHAShoot() error {
 	return nil
 }
 
-func (c *validationContext) validateScheduling(ctx context.Context, a admission.Attributes, authorizer authorizer.Authorizer, shootLister corelisters.ShootLister) error {
-	if a.GetOperation() == admission.Delete {
-		return nil
-	}
-
-	// If there is already a seedName in the shoot yaml, the scheduler will not create the Binding
-	// So we need to keep the validation here.
+func (c *validationContext) validateScheduling(ctx context.Context, a admission.Attributes, authorizer authorizer.Authorizer, shootLister corelisters.ShootLister, seedLister corelisters.SeedLister, shootStateLister corev1alpha1listers.ShootStateLister) error {
 	var (
 		shootIsBeingScheduled          = c.oldShoot.Spec.SeedName == nil && c.shoot.Spec.SeedName != nil
 		shootIsBeingRescheduled        = c.oldShoot.Spec.SeedName != nil && c.shoot.Spec.SeedName != nil && *c.shoot.Spec.SeedName != *c.oldShoot.Spec.SeedName
 		mustCheckSchedulingConstraints = shootIsBeingScheduled || shootIsBeingRescheduled
 	)
 
-	if shootIsBeingScheduled {
-		if a.GetOperation() == admission.Create {
+	switch a.GetOperation() {
+	case admission.Create:
+		if shootIsBeingScheduled {
 			if err := authorize(ctx, a, authorizer, "set .spec.seedName"); err != nil {
 				return err
 			}
 		}
+	case admission.Update:
+		if a.GetSubresource() == "binding" {
+			if c.shoot.DeletionTimestamp != nil {
+				return admission.NewForbidden(a, fmt.Errorf("shoot %s is being deleted, cannot be assigned to a seed", c.shoot.Name))
+			}
+
+			if c.oldShoot.Spec.SeedName != nil && c.shoot.Spec.SeedName == nil {
+				return admission.NewForbidden(a, fmt.Errorf("spec.seedName cannot be set to nil"))
+			}
 
+			if shootIsBeingRescheduled && !utilfeature.DefaultFeatureGate.Enabled(features.SeedChange) {
+				if err := apivalidation.ValidateImmutableField(c.oldShoot.Spec.SeedName, c.shoot.Spec.SeedName, field.NewPath("spec", "seedName")).ToAggregate(); err != nil {
+					return err
+				}
+			}
```

Comment:
There is one validation that was removed when the binding resource was initially added, but it is actually necessary now, since the shoot.spec can now be changed through the binding subresource:
https://github.com/gardener/gardener/blob/9aeae75a41f444d4f3b5173550e71630bc6f5404/plugin/pkg/shoot/validator/admission.go#L375-L379

Reply:
Hmm yes, I also remember that we had problems for this reason. But maybe it is ok for the control plane migration case (when the update is limited to just changing the seedName). I will double check a bit later today.

Reply:
The check that I mentioned should only be executed when the old seedName is not nil and the new seedName is not nil and they are different. I think it would not lead to any problems since w/e was supposed to be assigned to the shoot spec by the dns admission plugin should already have been assigned when the seedName was changed from nil, to not nil (shoot was assigned to a seed cluster).
Furthermore this would work as it did before, when there was no `binding` subresource. Back then, migrating the shoot would also cause the dns admission plugin to be called, but this did not lead to any problems.

Reply:
I have added the check. PTAL :)